File: haskell/list-ops/src/ListOps.hs
module ListOps
  ( length,
    reverse,
    map,
    filter,
    foldr,
    foldl',
    (++),
    concat,
  )
where

import Prelude hiding
  ( concat,
    filter,
    foldr,
    length,
    map,
    reverse,
    (++),
  )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ z [] = z
foldl' f z (x : xs) =
  let z' = z `f` x
   in seq z' $ foldl' f z' xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ z [] = z
foldr f z (x : xs) = x `f` foldr f z xs

length :: [a] -> Int
length [] = 0
length (_ : xs) = 1 + length xs

reverse :: [a] -> [a]
reverse [] = []
reverse (x : xs) = reverse xs ++ [x]

map :: (a -> b) -> [a] -> [b]
map _ [] = []
map f (x : xs) = f x : map f xs

filter :: (a -> Bool) -> [a] -> [a]
filter _ [] = []
filter p (x : xs)
  | p x = x : filter p xs
  | otherwise = filter p xs

(++) :: [a] -> [a] -> [a]
[] ++ ys = ys
(x : xs) ++ ys = x : xs ++ ys

concat :: [[a]] -> [a]
concat = foldr (++) []

File: typescript/list-ops/list-ops.ts
export class List<T> {

  private constructor(readonly values: T[]) { }

  public static create<T>(...values: T[]): List<T> {
    return new List(values)
  }

  forEach(cb: (item: any) => void): void {
    for (const item of this.values) {
      cb(item)
    }
  }

  append(list: List<T>): List<T> {
    return List.create(...this.values, ...list.values)
  }

  concat(list: List<List<T>>): List<T> {
    return list.foldl((acc: List<T>, curr: List<T>) => acc.append(curr), List.create(...this.values))
  }

  filter<T>(fn: ((item: T) => Boolean)): T[] {
    return this.foldr((acc: T[], curr: T) => fn(curr) ? [curr, ...acc] : acc, [])
  }

  length(): number {
    return this.foldr((acc: number, _: any) => acc + 1, 0)
  }

  map<T>(fn: ((item: any) => T)): T[] {
    return this.foldr((acc: T[], curr: any) => [fn(curr), ...acc], [])
  }

  foldl<T, R>(fn: ((acc: R, curr: T) => R), acc: R): R {
    this.forEach((item: T) => {
      acc = fn(acc, item)
    })

    return acc
  }

  foldr<T, R>(fn: ((acc: R, curr: T) => R), acc: R): R {
    const newList = List.create(...this.values.reverse())
    return newList.foldl(fn, acc)
  }

  reverse(): T[] {
    const newList = List.create(...this.values.reverse())
    return newList.values
  }
}

